### PR TITLE
Swappable plotting API: the `NDCube.plotter` property

### PR DESCRIPTION
**Pull request by Cadair**
This PR makes a new attribute on the NDCube object which provides all the plotting methods. This refactor allows people to easily bring their own plotting class, if say for instance someone wanted to build a plotting interface using Bokeh.

It works by defining a "Visualizer" class which takes one argument to the constructor, a reference to the parent `NDCube` object.

The new `NDCube.visualizer` descriptor, takes an object (uninitialised) when setting the attribute, the descriptor then intialises this object with a reference to the NDCube object it was assigned to, preventing people referencing the wrong instance of the Cube.

Here is a simple example:

```python
In [1]: import astropy.wcs
   ...: import numpy as np
   ...: 
   ...: from ndcube import NDCube
   ...: 
   ...: # Define data array.
   ...: data = np.random.rand(4, 4, 5)
   ...: # Define WCS transformations in an astropy WCS object.
   ...: wcs = astropy.wcs.WCS(naxis=3)
   ...: wcs.wcs.ctype = 'WAVE', 'HPLT-TAN', 'HPLN-TAN'
   ...: wcs.wcs.cunit = 'Angstrom', 'deg', 'deg'
   ...: wcs.wcs.cdelt = 0.2, 0.5, 0.4
   ...: wcs.wcs.crpix = 0, 2, 2
   ...: wcs.wcs.crval = 10, 0.5, 1
   ...: wcs.wcs.cname = 'wavelength', 'HPC lat', 'HPC lon'
   ...: 
   ...: # Now instantiate the NDCube
   ...: my_cube = NDCube(data, wcs=wcs)

In [2]: my_cube._vis_instance

In [3]: my_cube.visualizer
Out[3]: <ndcube.visualization.mpl_visualizer.MatplotlibVisualizer at 0x7f7bbdb20220>

In [4]: my_cube._vis_instance
Out[4]: <ndcube.visualization.mpl_visualizer.MatplotlibVisualizer at 0x7f7bbdb20220>

In [6]: from ndcube.visualization.mpl_visualizer import MatplotlibVisualizer

In [7]: class TestVis(MatplotlibVisualizer):
   ...:     pass
   ...: 

In [8]: my_cube.visualizer = TestVis

In [9]: my_cube.visualizer
Out[9]: <__main__.TestVis at 0x7f7c2c57b880>

In [10]: my_cube.visualizer = TestVis(my_cube)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-22bc892c7bdd> in <module>
----> 1 my_cube.visualizer = TestVis(my_cube)

~/Git/ndcube/ndcube/ndcube.py in __set__(self, obj, value)
    597     def __set__(self, obj, value):
    598         if not isinstance(value, type):
--> 599             raise TypeError("NDCube.visualizer can only be set with an uninitialised visualizer instance")
    600         obj._vis_instance = value(obj)
    601 

TypeError: NDCube.visualizer can only be set with an uninitialised visualizer instance

In [11]: my_cube.visualizer.plot()
Out[11]: <sunpy.visualization.animator.wcs.ArrayAnimatorWCS at 0x7f7bbe63a790>
```

## Questions that need answering:

* Should we remove `NDCube.plot`?
* Should we define some ABC for visualizer instances, if so what should it specify?
* What to do about `_as_mpl_axis`?

### CODE REVIEW

**Comment by DanRyanIrish** [comment]
@Cadair, this is great.  Regarding your questions:
 
>     * Should we remove `NDCube.plot`?

I say yes.  2.0 is a create chance to do this.  For now we can just have `NDCube.visualizer.plot`.  Further methods, like `NDCube.visualizer.step`, can be added after 2.0.

>     * Should we define some ABC for visualizer instances, if so what should it specify?

From a previous converstation, there should be an ABC but it should be minimalist giving other users/developers maximum freedom on how to develop their visualization tool.  So the ABC should probably only define the descriptor.

>     * What to do about `_as_mpl_axis`?

Not sure what the issue is here.  So not sure about the solution

>  * Suggestions on the name welcome.

What about `NDCube.viewer` or `NDCube.artist`?  `visualizer` is a bit hard to say and I can see us tripping over it in tutorials.

**Comment by Cadair** [comment]
> `NDCube.viewer`,  `NDCube.artist`, `NDCube.plotter`, `NDCube.gallery`, `NDCube.canvas`

The only one of these I really like is `plotter`.

**Comment by DanRyanIrish** [comment]
> The only one of these I really like is `plotter`.

I'm OK with that.

**Comment by dstansby** [comment]
I like plotter too -it's what `pyvista` use as their equivalent of a matplotlib Figure.

**Comment by DanRyanIrish** [comment]
> > ```
> > * Should we remove `NDCube.plot`?
> > ```
> 
> I say yes. 2.0 is a create chance to do this. For now we can just have `NDCube.visualizer.plot`. Further methods, like `NDCube.visualizer.step`, can be added after 2.0.

Subsequent to the sunpy community meeting yesterday, I think the concensus was that this implementation is a good idea but that we should keep an `NDCube.plot` method to points to `NDCube.plotter.plot` if such a method exists.

> > * Suggestions on the name welcome.
> 
> `visualizer` is a bit hard to say and I can see us tripping over it in tutorials. Here are a few alternatives I brainstormed. `NDCube.viewer`, `NDCube.artist`, `NDCube.plotter`, `NDCube.gallery`, `NDCube.canvas`

I also think the concensus is that we change `Visualizer` to `Plotter`

**Comment by Cadair** [comment]
So I updated some things.

1) I moved some of our tests around as they were annoying me.
2) I "solved" the issue with "_as_mpl_axes", thoughts on the solution very welcome.
3) I re-added `plot()`.
4) Renamed it all "plotter"
5) Added an ABC for good measure.

**Review by DanRyanIrish — COMMENTED** [review]
@Cadair, this looks good so far.  I've left a couple of comments, more to help with my understanding.

From the initial description of the behaviour of the `plotter` attribute, it's possible change the `plotter` class on a single `NDCube` as many times as you like simply by doing `my_cube.plotter = MyPlotter`.  Is that what this code is facilitating now?

File: ndcube/visualization/mpl_plotter.py [hunk 1]
```diff
@@ -7,15 +7,19 @@
...
 
-class NDCubePlotMixin:
+class MatplotlibPlotter(BasePlotter):
     """
-    Add plotting functionality to a NDCube class.
+    Provide visualization methods for NDCube which use `matplotlib`.
     """
 
+    def __init__(self, ndcube):
+        self._ndcube = ndcube
```
[the remark this answers is not in the document]
Comment (Cadair) [reply ndcube/visualization/mpl_plotter.py:21]: Yeah I think I can just drop this, I added the ABC and forgot to update this class.

File: ndcube/visualization/descriptor.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+class PlotterDescriptor:
+    def __init__(self, default_type=None):
+        self._default_type = default_type
+
+    def __set_name__(self, owner, name):
```
[the remark this answers is not in the document]
Comment (Cadair) [reply ndcube/visualization/descriptor.py:5]: This is a little odd. This is called at the point where the class with the descriptor attached is defined.

So in this specific case this means when the `NDCube` class is imported for the first time, this method is called after the descriptor is instantiated, to tell the descriptor the name of the attribute that refers to the descriptor on it's parent class (NDCube). So in this case `name` should always be `'plotter'` as that's the attribute name on the `NDCube` instance.

**Comment by Cadair** [comment]
> From the initial description of the behaviour of the `plotter` attribute, it's possible change the `plotter` class on a single `NDCube` as many times as you like simply by doing `my_cube.plotter = MyPlotter`. Is that what this code is facilitating now?

Correct, it's like setting a property. (one with a `@setter`).

File: ndcube/visualization/descriptor.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DanRyanIrish) [reply ndcube/visualization/descriptor.py:5]: OK, that makes sense.  But only after reading it a few times!  But that seems to be more to do with the fact that the workflow is pretty complicated!

File: ndcube/visualization/tests/test_sequence_plotting.py [hunk 3]
```diff
@@ -0,0 +1,778 @@
+# # -*- coding: utf-8 -*-
```
Comment (DanRyanIrish) [inline ndcube/visualization/tests/test_sequence_plotting.py:1]: Looks like this file should be deleted?

**Comment by DanRyanIrish** [comment]
@Cadair this looks fine except for my comment on `test_sequence_plotting.py`.

**Review by DanRyanIrish — APPROVED** [review]

**Pull request merged**